packages/mui-lab/src/Masonry/Masonry.js

```diff
@@ -182,80 +182,106 @@ const Masonry = React.forwardRef(function Masonry(inProps, ref) {
 
   const classes = useUtilityClasses(ownerState);
 
-  React.useEffect(() => {
-    const handleResize = () => {
-      const parentWidth = masonryRef.current.clientWidth;
-      const childWidth = masonryRef.current.firstChild.clientWidth;
-      const firstChildComputedStyle = window.getComputedStyle(masonryRef.current.firstChild);
-      const firstChildMarginLeft = parseToNumber(firstChildComputedStyle.marginLeft);
-      const firstChildMarginRight = parseToNumber(firstChildComputedStyle.marginRight);
-
-      if (parentWidth === 0 || childWidth === 0) {
-        return;
-      }
+  const observer = React.useRef(
+    typeof ResizeObserver === 'undefined'
+      ? undefined
+      : new ResizeObserver((elements) => {
+          if (!elements) {
+            return;
+          }
+          let masonry;
+          let masonryFirstChild;
+          let parentWidth;
+          let childWidth;
+          if (elements[0].target.className.includes(classes.root)) {
+            masonry = elements[0].target;
+            parentWidth = elements[0].contentRect.width;
+            masonryFirstChild = elements[1]?.target || masonryRef.current.firstChild;
+            childWidth =
+              masonryFirstChild?.contentRect?.width || masonryFirstChild?.clientWidth || 0;
+          } else {
+            masonryFirstChild = elements[0].target;
+            childWidth = elements[0].contentRect.width;
+            masonry = elements[1]?.target || masonryRef.current;
```

Comment:
```suggestion
            masonry = elements[1]?.target || masonryFirstChild.parentElement;
```

Double check the API